components/helm-broker/internal/controller/addonsconfiguration_controller.go

```diff
@@ -0,0 +1,500 @@
+	return &AddonsConfigurationController{reconciler: reconciler}
+}
+
+// Start starts a controller
+func (acc *AddonsConfigurationController) Start(mgr manager.Manager) error {
+	// Create a new controller
+	c, err := controller.New("addonsconfiguration-controller", mgr, controller.Options{Reconciler: acc.reconciler})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to AddonsConfiguration
+	err = c.Watch(&source.Kind{Type: &addonsv1alpha1.AddonsConfiguration{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileAddonsConfiguration{}
+
+// ReconcileAddonsConfiguration reconciles a AddonsConfiguration object
+type ReconcileAddonsConfiguration struct {
+	log logrus.FieldLogger
+	client.Client
+	scheme *runtime.Scheme
+
+	chartStorage  chartStorage
+	bundleStorage bundleStorage
+
+	brokerFacade      brokerFacade
+	brokerSyncer      brokerSyncer
+	docsTopicProvider docsProvider
+
+	protection     protection
+	bundleProvider bundleProvider
+
+	// syncBroker informs ServiceBroker should be resync, it should be true if
+	// operation insert/delete was made on storage
+	syncBroker  bool
+	developMode bool
+}
+
+// NewReconcileAddonsConfiguration returns a new reconcile.Reconciler
+func NewReconcileAddonsConfiguration(mgr manager.Manager, bp bundleProvider, brokerFacade brokerFacade, chartStorage chartStorage, bundleStorage bundleStorage, developMode bool, docsTopicProvider docsProvider, brokerSyncer brokerSyncer) reconcile.Reconciler {
+	return &ReconcileAddonsConfiguration{
+		log:    logrus.WithField("controller", "addons-configuration"),
+		Client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+
+		chartStorage:  chartStorage,
+		bundleStorage: bundleStorage,
+
+		bundleProvider: bp,
+		protection:     protection{},
+
+		brokerSyncer:      brokerSyncer,
+		brokerFacade:      brokerFacade,
+		docsTopicProvider: docsTopicProvider,
+
+		developMode: developMode,
+		syncBroker:  false,
+	}
+}
+
+// Reconcile reads that state of the cluster for a AddonsConfiguration object and makes changes based on the state read
+// and what is in the AddonsConfiguration.Spec
+func (r *ReconcileAddonsConfiguration) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	instance := &addonsv1alpha1.AddonsConfiguration{}
+	err := r.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if instance.DeletionTimestamp != nil {
+		if err := r.deleteAddonsProcess(instance); err != nil {
+			return reconcile.Result{RequeueAfter: time.Second * 15}, exerr.Wrapf(err, "while deleting AddonConfiguration %q", request.NamespacedName)
+		}
+		return reconcile.Result{}, nil
+	}
+
+	if instance.Status.ObservedGeneration == 0 {
+		r.log.Infof("Start add AddonsConfiguration %s/%s process", instance.Name, instance.Namespace)
+		pendingInstance, err := r.setPendingStatus(instance)
```

Comment:
after consultation with @mszostok pending state is not necessary, I am not sure we should implement it

Reply:
I won't delete the pending status step. I think it may be helpful for debugging if the controller logic will have some bugs. I merged it to one step - adding finalizer and setting pending status.